### ISSUE TITLE
bug:  apple-framework-CoreFoundation-11.0.0: used as improper sort of dependency

### ISSUE THREAD

**Issue by tartavull** [nix-community/home-manager#5368]
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

When I install a package, for example, `buf` I get the following error `apple-framework-CoreFoundation-11.0.0: used as improper sort of dependency` but when I run `nix-shell -p buf` it works perfectly.

This is also true for many other packages.

```
home-manager switch

these 7 derivations will be built:
  /nix/store/88s905wwcv4j4ac5lp3v2bpp3qdfa3l4-buf-1.27.0.drv
  /nix/store/7a7njmckwa97cmw10aw45aa5k6bjcscq-home-manager-path.drv
  /nix/store/bynzz5qvd3mjczs4lr4334fkw8ffwlg0-home-manager-fonts.drv
  /nix/store/c63y72nmzr7bcwihhn5bqki80alm80nz-activation-script.drv
  /nix/store/ypxzyrhbiz7q95nblf3g8pq1gk4gasw0-home-manager-applications.drv
  /nix/store/cgqylcggx5dca6wx4lhaa7sxzgnk0mzw-home-manager-files.drv
  /nix/store/dh31lbdy81ivh3r5vl871m5b6wnf1ypj-home-manager-generation.drv
building '/nix/store/88s905wwcv4j4ac5lp3v2bpp3qdfa3l4-buf-1.27.0.drv'...
Running phase: unpackPhase
unpacking source archive /nix/store/j5ysaxx453mm259fsb3c38fadbrmvssk-source
source root is source
Running phase: patchPhase
applying patch /nix/store/r8wr0f85br2z0vq4gmh0vzkgyx1ny5a8-skip_test_requiring_network.patch
patching file private/bufpkg/buftesting/buftesting.go
Hunk #1 succeeded at 104 (offset 5 lines).
Running phase: updateAutotoolsGnuConfigScriptsPhase
Running phase: configurePhase
apple-framework-CoreFoundation-11.0.0: used as improper sort of dependency
error: builder for '/nix/store/88s905wwcv4j4ac5lp3v2bpp3qdfa3l4-buf-1.27.0.drv' failed with exit code 1;
       last 10 log lines:
       > Running phase: unpackPhase
       > unpacking source archive /nix/store/j5ysaxx453mm259fsb3c38fadbrmvssk-source
       > source root is source
       > Running phase: patchPhase
       > applying patch /nix/store/r8wr0f85br2z0vq4gmh0vzkgyx1ny5a8-skip_test_requiring_network.patch
       > patching file private/bufpkg/buftesting/buftesting.go
       > Hunk #1 succeeded at 104 (offset 5 lines).
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > apple-framework-CoreFoundation-11.0.0: used as improper sort of dependency
       For full logs, run 'nix log /nix/store/88s905wwcv4j4ac5lp3v2bpp3qdfa3l4-buf-1.27.0.drv'.
error: 1 dependencies of derivation '/nix/store/ypxzyrhbiz7q95nblf3g8pq1gk4gasw0-home-manager-applications.drv' failed to build
error: 1 dependencies of derivation '/nix/store/bynzz5qvd3mjczs4lr4334fkw8ffwlg0-home-manager-fonts.drv' failed to build
error: 1 dependencies of derivation '/nix/store/7a7njmckwa97cmw10aw45aa5k6bjcscq-home-manager-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/dh31lbdy81ivh3r5vl871m5b6wnf1ypj-home-manager-generation.drv' failed to build
```
```
nix-shell -p buf
this path will be fetched (12.12 MiB download, 60.00 MiB unpacked):
  /nix/store/svika1rr2w3axzlv0g5pzd0abgnd0nc9-buf-1.27.0
copying path '/nix/store/svika1rr2w3axzlv0g5pzd0abgnd0nc9-buf-1.27.0' from 'https://cache.nixos.org'...
```

As far as I can tell I'm using the same version for both home-manager and nixpkgs
```
nix-channel --list

home-manager https://github.com/nix-community/home-manager/archive/release-23.11.tar.gz
nixpkgs https://nixos.org/channels/nixpkgs-23.11-darwin
```

### Maintainer CC

not sure.

### System information

```markdown
- system: `"aarch64-darwin"`
 - host os: `Darwin 23.4.0, macOS 14.4.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.12.0`
 - channels(ignaciotartavull): `"home-manager-23.11.tar.gz, nixpkgs-23.11-darwin"`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/Users/ignaciotartavull/.nix-defexpr/channels/nixpkgs`
```

**Comment by tartavull** [comment]
I'm an idiot, ignored.

**Comment by nerosnm** [comment]
I'm seeing this kind of error message when running `home-manager build`, but as far as I can tell I'm not using any such kind of go overlay. Can you point me in the right direction? This is the only issue I could find that mentioned this "improper sort of dependency" error.

**Comment by nerosnm** [comment]
Ok, I figured it out. I was overriding the `nix` package to pull from `nixpkgs-master`, and it seems I also needed to pull `nixos-option` from there. Otherwise, `nix` (a dependency of `nixos-option`, a dependency of `home-manager`) would fail to build with `apple-framework-CoreFoundation-11.0.0: used as improper sort of dependency`.

This all stemmed from my activation of `programs.home-manager.enable = true;`, in case that helps anyone else find this in future.
